projects/packages/sync/src/class-lock.php

```diff
@@ -42,18 +42,37 @@ class Lock {
 	 * @return boolean True if succeeded, false otherwise.
 	 */
 	public function attempt( $name, $expiry = self::LOCK_TRANSIENT_EXPIRY ) {
-		$name        = self::LOCK_PREFIX . $name;
-		$locked_time = get_option( $name );
+
+		global $wpdb;
+		$name = self::LOCK_PREFIX . $name;
+
+		// Options API is explicitly bypassed. This is because with high concurrency request to the
+		// site multiple processes would pass checks and send concurrent data to WP.
+
+		$row         = $wpdb->get_row( $wpdb->prepare( "SELECT option_value FROM $wpdb->options WHERE option_name = %s LIMIT 1", $name ) );
+		$locked_time = false;
+		if ( is_object( $row ) ) {
+			$locked_time = $row->option_value;
+		}
 
 		if ( $locked_time ) {
 			if ( microtime( true ) < $locked_time ) {
 				return false;
+			} else {
+				// If expired delete but don't send. That will occurr in new request to avoid race conditions.
+				$wpdb->delete( $wpdb->options, array( 'option_name' => $name ) );
+				return false;
 			}
 		}
 		$locked_time = microtime( true ) + $expiry;
-		update_option( $name, $locked_time );
+		$result      = $wpdb->query( $wpdb->prepare( "INSERT INTO `$wpdb->options` (`option_name`, `option_value`, `autoload` ) VALUES (%s, %s, %s)", $name, maybe_serialize( $locked_time ), 'no' ) );
```

Comment:
Why `maybe_serialize`? I'd expect `maybe_unserialize` on every read then?

Reply:
this came from the core options API usage. no longer applies but you are correct it should always match.